### PR TITLE
Improve documentation of SoftmaxOutput

### PR DESCRIPTION
**Pull request by jens-mueller-sociomantic**
This change improves the documentation of the `SoftmaxOutput` symbol by
adding information regarding the calculated cross entropy loss. The old
documentation only contained information on the softmax function. The
documentation was also wrong in the sense that the softmax function has
no label information as part of its input. It only turns a vector into a
probability vector. The changed documentation explains the softmax
function and the cross entropy loss on their own and their combination
to describe the calculation represented by the `SoftmaxOutput` symbol.

### CODE REVIEW

**Comment by jens-mueller-sociomantic** [comment]
Changes based on some offline feedback by @alaksiej-stankievicx-sociomantic.

```diff
diff --git a/src/mxnet/Symbol.d b/src/mxnet/Symbol.d
index 9c25d44..12572a6 100644
--- a/src/mxnet/Symbol.d
+++ b/src/mxnet/Symbol.d
@@ -736,29 +736,32 @@ public class SoftmaxOutput : Symbol
 {
     /***************************************************************************

-        Constructs a symbol representing the cross entropy loss of the label(s)
-        as a predictions vector and the softmax of input(s)
+        Constructs a symbol representing the loss calculated as cross entropy
+        between label(s) and softmax of input(s)

         The cross entropy measures the similarity of two probability
         distributions over the same set of events. By computing the `softmax`
         of an input vector and the one hot encoding of a label the inputs are
-        represented as probability vectors. The cross entropy H(y, x) between
-        these discrete distributions is then calculated by `-y' *
-        log(softmax(x))` where `'` denotes transposition of a one hot encoded
-        label vector y and `*` denotes the dot product.
+        represented as probability vectors. The cross entropy `H(y, x)` between
+        these discrete distributions is then calculated by
+        `-y' * log(softmax(x))` where `'` denotes transposition of a one hot
+        encoded label vector `y`.

-        The softmax function is defined by `softmax(x) = exp(x) / sum exp(x)`.
+        The softmax function is defined by `softmax(x) = exp(x) / sum(exp(x))`.
         It transforms its input vector into a probability vector. The softmax
         function is a generalization of the logistic function allowing for more
         than two classes.

-        This function sums the cross entropy losses over all label-input pairs,
-        formally `-sum_i y_i' * log(softmax(x_i))` where y_i and x_i have same
-        length, that is, the number of label classes.
+        In case you have more than label-input pair, this function sums the
+        cross entropy losses over all pairs, formally
+        `-sum(y_i' * log(softmax(x_i)))` where the summation runs over the
+        index `i` and each vectors `y_i` and `x_i` have same length, that is,
+        the number of label classes.

-        In a forward pass only the softmax is calculated, turning the input
-        into a probability vector. Additionally, in a backward pass the cross
-        entropy is calculated including gradient information.
+        For prediction purposes (i.e., on a forward pass) only the softmax is
+        calculated, turning the input into a probability vector. Additionally,
+        in training (i.e., on a backward pass) the cross entropy is used for
+        calculating loss and gradient information.

         Params:
             input = input symbol to apply softmax to
```

File: src/mxnet/Symbol.d [hunk 1]
```diff
@@ -736,14 +736,36 @@ public class SoftmaxOutput : Symbol
...
-        Constructs a symbol representing the result of performing the softmax
-        on `input` and `label`.
-
-        The softmax function is defined as `exp(input_label) / sum(exp(input))`.
+        Constructs a symbol representing the loss calculated as cross entropy
+        between label(s) and softmax of input(s)
+
+        The cross entropy measures the similarity of two probability
+        distributions over the same set of events. By computing the `softmax`
+        of an input vector and the one hot encoding of a label the inputs are
```
Comment (alaksiej-stankievicx-dunnhumby) [inline src/mxnet/Symbol.d]: the last `inputs` here sounds a bit ambiguous (I understand what you mean, but it could expressed better IMO)

Comment (alaksiej-stankievicx-dunnhumby) [reply src/mxnet/Symbol.d]: an offer: `they both are represented...`

File: src/mxnet/Symbol.d [hunk 2]
```diff
@@ -736,14 +736,36 @@ public class SoftmaxOutput : Symbol
...
+        these discrete distributions is then calculated by
+        `-y' * log(softmax(x))` where `'` denotes transposition of a one hot
+        encoded label vector `y`.
+
+        The softmax function is defined by `softmax(x) = exp(x) / sum(exp(x))`.
+        It transforms its input vector into a probability vector. The softmax
+        function is a generalization of the logistic function allowing for more
+        than two classes.
+
+        In case you have more than label-input pair, this function sums the
```
Comment (alaksiej-stankievicx-dunnhumby) [inline src/mxnet/Symbol.d]: `more than one`

File: src/mxnet/Symbol.d (hunk 1 above)
Comment (jens-mueller-sociomantic) [reply src/mxnet/Symbol.d]: True. How about `both are represented as probability vectors`?

**Comment by jens-mueller-sociomantic** [comment]
```diff
diff --git a/src/mxnet/Symbol.d b/src/mxnet/Symbol.d
index 12572a6..ef8182b 100644
--- a/src/mxnet/Symbol.d
+++ b/src/mxnet/Symbol.d
@@ -739,24 +739,24 @@ public class SoftmaxOutput : Symbol
         Constructs a symbol representing the loss calculated as cross entropy
         between label(s) and softmax of input(s)

+        A label is either already a probability vector or treated internally as
+        such by one-hot encoding.
+
+        An input is transformed to a probability vector by the softmax
+        function which is defined as `softmax(x) = exp(x) / sum(exp(x))`. The
+        softmax function is a generalization of the logistic function allowing
+        for more than two classes.
+
         The cross entropy measures the similarity of two probability
-        distributions over the same set of events. By computing the `softmax`
-        of an input vector and the one hot encoding of a label the inputs are
-        represented as probability vectors. The cross entropy `H(y, x)` between
-        these discrete distributions is then calculated by
-        `-y' * log(softmax(x))` where `'` denotes transposition of a one hot
-        encoded label vector `y`.
-
-        The softmax function is defined by `softmax(x) = exp(x) / sum(exp(x))`.
-        It transforms its input vector into a probability vector. The softmax
-        function is a generalization of the logistic function allowing for more
-        than two classes.
-
-        In case you have more than label-input pair, this function sums the
+        distributions over the same set of events. It is calculated between two
+        probability vectors as `-y' * log(softmax(x))` where `'` denotes
+        transposition.
+
+        In case you have more than one label-input pair, this function sums the
         cross entropy losses over all pairs, formally
         `-sum(y_i' * log(softmax(x_i)))` where the summation runs over the
-        index `i` and each vectors `y_i` and `x_i` have same length, that is,
-        the number of label classes.
+        index `i` and all vectors `y_i` and `x_i` have the same length, that
+        is, the number of label classes.

         For prediction purposes (i.e., on a forward pass) only the softmax is
         calculated, turning the input into a probability vector. Additionally,
```

**Comment by jens-mueller-sociomantic** [comment]
```diff
diff --git a/src/mxnet/Symbol.d b/src/mxnet/Symbol.d
index ef8182b..b7a2835 100644
--- a/src/mxnet/Symbol.d
+++ b/src/mxnet/Symbol.d
@@ -759,13 +759,15 @@ public class SoftmaxOutput : Symbol
         is, the number of label classes.

         For prediction purposes (i.e., on a forward pass) only the softmax is
-        calculated, turning the input into a probability vector. Additionally,
-        in training (i.e., on a backward pass) the cross entropy is used for
-        calculating loss and gradient information.
+        calculated, turning the input into a probability vector. In training
+        (i.e., a forward pass with objective followed by a backward pass) the
+        cross entropy is used for calculating objective and gradient
+        information.

         Params:
             input = input symbol to apply softmax to
-            label = label symbol to apply cross entropy to
+            label = reference for calculating the cross entropy of softmaxed
+                    input
             normalization = normalization applied to the gradient; defaults to
                             batch
             grad_scale = scale factor for scaling the gradient; defaults to 1
```

**Comment by jens-mueller-sociomantic** [comment]
```diff
diff --git a/src/mxnet/Symbol.d b/src/mxnet/Symbol.d
index b7a2835..feff563 100644
--- a/src/mxnet/Symbol.d
+++ b/src/mxnet/Symbol.d
@@ -720,7 +720,36 @@ public enum SoftmaxOutputNormalization

 /*******************************************************************************

-    A symbol representing softmax (and cross entropy loss)
+    A symbol representing softmax (and cross entropy loss for calculating
+    objective and gradient information)
+
+    This symbol represents the application of the softmax function to the
+    input. For a backward pass the cross entropy is calculated using the label
+    as reference for obtaining objective and gradient information.
+
+    A label is either already a probability vector or treated internally as
+    such by one-hot encoding.
+
+    An input is transformed to a probability vector by the softmax function
+    which is defined as `softmax(x) = exp(x) / sum(exp(x))`. The softmax
+    function is a generalization of the logistic function allowing for more
+    than two classes.
+
+    The cross entropy measures the similarity of two probability distributions
+    over the same set of events. It is calculated between two probability
+    vectors as `-y' * log(softmax(x))` where `'` denotes transposition. It is
+    used as an objective function in this symbol.
+
+    In case you have more than one label-input pair, this function sums the
+    cross entropy losses over all pairs, formally
+    `-sum(y_i' * log(softmax(x_i)))` where the summation runs over the index
+    `i` and all vectors `y_i` and `x_i` have the same length, that is, the
+    number of label classes.
+
+    For prediction purposes (i.e., on a forward pass) only the softmax is
+    calculated, turning the input into a probability vector. In training (i.e.,
+    a forward pass with objective followed by a backward pass) the cross
+    entropy is used for calculating objective and gradient information.

     An SoftmaxOutput allocates MXNet resources accessed through an
     `MXNetHandle`. Its resources should be freed by calling `freeHandle` when
@@ -736,33 +765,7 @@ public class SoftmaxOutput : Symbol
 {
     /***************************************************************************

-        Constructs a symbol representing the loss calculated as cross entropy
-        between label(s) and softmax of input(s)
-
-        A label is either already a probability vector or treated internally as
-        such by one-hot encoding.
-
-        An input is transformed to a probability vector by the softmax
-        function which is defined as `softmax(x) = exp(x) / sum(exp(x))`. The
-        softmax function is a generalization of the logistic function allowing
-        for more than two classes.
-
-        The cross entropy measures the similarity of two probability
-        distributions over the same set of events. It is calculated between two
-        probability vectors as `-y' * log(softmax(x))` where `'` denotes
-        transposition.
-
-        In case you have more than one label-input pair, this function sums the
-        cross entropy losses over all pairs, formally
-        `-sum(y_i' * log(softmax(x_i)))` where the summation runs over the
-        index `i` and all vectors `y_i` and `x_i` have the same length, that
-        is, the number of label classes.
-
-        For prediction purposes (i.e., on a forward pass) only the softmax is
-        calculated, turning the input into a probability vector. In training
-        (i.e., a forward pass with objective followed by a backward pass) the
-        cross entropy is used for calculating objective and gradient
-        information.
+        Constructs `SoftmaxOutput` symbol

         Params:
             input = input symbol to apply softmax to
```

File: src/mxnet/Symbol.d [hunk 3]
```diff
@@ -736,14 +765,12 @@ public class SoftmaxOutput : Symbol
...
-        on `input` and `label`.
-
-        The softmax function is defined as `exp(input_label) / sum(exp(input))`.
+        Constructs `SoftmaxOutput` symbol
 
         Params:
             input = input symbol to apply softmax to
-            label = label symbol to apply softmax to
+            label = reference for calculating the cross entropy of softmaxed
+                    input
```
Comment (joseph-wakeling-sociomantic) [inline src/mxnet/Symbol.d]: I don't understand this documentation at all.  What does "Reference for calculating the cross entropy" mean?

File: src/mxnet/Symbol.d [hunk 4]
```diff
@@ -720,7 +720,36 @@ public enum SoftmaxOutputNormalization
...
-    A symbol representing softmax (and cross entropy loss)
+    A symbol representing softmax (and cross entropy loss for calculating
+    objective and gradient information)
+
+    This symbol represents the application of the softmax function to the
+    input. For a backward pass the cross entropy is calculated using the label
+    as reference for obtaining objective and gradient information.
+
+    A label is either already a probability vector or treated internally as
+    such by one-hot encoding.
```
Comment (joseph-wakeling-sociomantic) [inline src/mxnet/Symbol.d:731]: Can you clarify what you mean here, please?

File: src/mxnet/Symbol.d (hunk 3 above)
Comment (jens-mueller-sociomantic) [reply src/mxnet/Symbol.d]: Would `ground truth for calculating the cross entropy` be any better?

File: src/mxnet/Symbol.d (hunk 4 above)
Comment (jens-mueller-sociomantic) [reply src/mxnet/Symbol.d:731]: What is it that you don't understand? `probability vector`? `one-hot encoding`? Or how a label can be treated as a one-hot encoded vector?

File: src/mxnet/Symbol.d (hunk 3 above)
Comment (joseph-wakeling-sociomantic) [reply src/mxnet/Symbol.d]: It's a little better, but I think you can still be clearer about what it is (or can be).

Remember, this documentation should be about trying to help me understand what I need to provide to the constructor, as a developer using this library.  It's important to be as concrete and clear as possible to avoid misunderstanding.

Comment (jens-mueller-sociomantic) [reply src/mxnet/Symbol.d]: How about this phrasing `symbol representing the ground truth label information for calculating the cross entropy together with the softmaxed input`?

File: src/mxnet/Symbol.d (hunk 4 above)
Comment (joseph-wakeling-sociomantic) [reply src/mxnet/Symbol.d:731]: The `one-hot encoding` is new to me.  But I think more importantly, it's not clear to me what either option means in terms of how that impacts the calculation being done.

File: src/mxnet/Symbol.d (hunk 3 above)
Comment (joseph-wakeling-sociomantic) [reply src/mxnet/Symbol.d]: TBH that feels like adding words without clarifying meaning :-\

Put it this way: how would you explain what this symbol means to your grandparents (assuming they are not machine learning experts themselves)?

Comment (jens-mueller-sociomantic) [reply src/mxnet/Symbol.d]: That is why we have expanded the documentation of the class. As it explains what cross entropy is and should allow you to understand what this parameter is about.

File: src/mxnet/Symbol.d (hunk 4 above)
Comment (jens-mueller-sociomantic) [reply src/mxnet/Symbol.d:731]: Conceptually the calculations are the same. It is just viewing a label as a probability vector.

File: src/mxnet/Symbol.d (hunk 3 above)
Comment (joseph-wakeling-sociomantic) [reply src/mxnet/Symbol.d]: Yea, but I'd rather have that information clearly (and minimally) written here too.  Ideally the constructor parameters doc should be reasonably comprehensible without having to read the class doc in too much depth.

Comment (jens-mueller-sociomantic) [reply src/mxnet/Symbol.d]: I don't know how to condense all the background for the parameter. Do you have a suggestion? That might also help me understand what is not clear enough yet.

**Comment by jens-mueller-sociomantic** [comment]
@joseph-wakeling-sociomantic I pushed a suggestion for the documentation of the label parameter.

File: src/mxnet/Symbol.d [hunk 5]
```diff
@@ -769,8 +769,7 @@ public class SoftmaxOutput : Symbol
 
         Params:
             input = input symbol to apply softmax to
-            label = reference for calculating the cross entropy of softmaxed
-                    input
+            label = ground truth for the softmaxed input to compare against
```
Comment (joseph-wakeling-sociomantic) [inline src/mxnet/Symbol.d]: By `softmaxed input`, do you mean the result of applying softmax to the input?

Isn't this perhaps better described as the _output_ of the `softmax` function?

**Review by joseph-wakeling-sociomantic — APPROVED** [review]
LGTM, please squash and we can take this in :-)

**Comment by jens-mueller-sociomantic** [comment]
@joseph-wakeling-sociomantic Still good?

**Review by alaksiej-stankievicx-dunnhumby — APPROVED** [review]
